src/main/java/org/mockito/internal/configuration/plugins/PluginRegistry.java

```diff
@@ -50,10 +59,36 @@ MockMaker getMockMaker() {
      * Returns the instantiator provider available for the current runtime.
      *
      * <p>Returns {@link org.mockito.internal.creation.instance.DefaultInstantiatorProvider} if no
-     * {@link org.mockito.plugins.InstantiatorProvider} extension exists or is visible in the current classpath.</p>
+     * {@link org.mockito.plugins.InstantiatorProvider2} extension exists or is visible in the
+     * current classpath.</p>
      */
-    InstantiatorProvider getInstantiatorProvider() {
-      return instantiatorProvider;
+    InstantiatorProvider2 getInstantiatorProvider() {
+        // Lazily init defaultInstantiatorProvider as all dependencies of the default instantiator
+        // provider might not be available when this class is initialized
+        if (defaultInstantiatorProvider == null) {
+            defaultInstantiatorProvider = (new DefaultMockitoPlugins()).getDefaultPlugin(InstantiatorProvider.class);
+        }
+
+        if (instantiatorProvider.getClass() != defaultInstantiatorProvider.getClass()) {
```

Comment:
Let's keep the complexity of loading plugins in the ```PluginLoader``` type. Otherwise it is hard to understand what's going on here. Example implementation:

```
Object plugin = new PluginLoader(pluginSwitch).loadPlugin(InstantiatorProvider.class, InstantiatorProvider2.class);
//returned plugin is either of those 2 types, we can use 'instanceof' to find out
```

then we can create a new type that adapts to the ```InstantiatorProvider2``` interface so that PluginRegistry could look like this:

```
private final InstantiatorProvider2 instantiatorProvider2 = 
        new InstantiatorProviderWrapper(
            new PluginLoader(pluginSwitch).loadPlugin(InstantiatorProvider.class, InstantiatorProvider2.class));
```

I think that implementing above will remove the need for ```LegacyDefaultInstantiatorProvider``` class included in this PR.

Reply:
Good suggestion, done.

I still have to keep legacy support in DefaultMockitoPlugins#getDefaultPlugin though.

I factored the compatibility wrappers InstantiatorProvider2 <-> InstantiatorProvider into separate classes.

Reply:
Thank you! I like the new types a lot.